cmd/log/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+        let last_char = buffer[last_idx];
+        if last_char != 0 && last_char != b'\r' {
+            let mut total_count = 0;
+            for (char_count, (i, _line)) in
+                zip(char_counts, lines.clone().iter().enumerate())
+            {
+                if total_count <= last_idx
+                    && total_count + char_count > last_idx
+                {
+                    lines.swap_remove(i);
+                    //println!("removed partial line: {}", line);
+                    break;
+                }
+                total_count += char_count;
+            }
+        }
+
+        for (i, l) in lines.iter().enumerate() {
+            //
+            // extract and verify timestamp
+            //
+            let timestamp = extract_timestamp(l);
+            if let Err(_err) = timestamp {
+                //println!("invalid Log Entry found: {:?}", l);
+                continue;
+            }
+            let timestamp: u64 = timestamp.unwrap();
+
+            //
+            // Insert line into map
+            //
+            if let std::collections::btree_map::Entry::Vacant(_e) =
+                full_log.entry(timestamp)
+            {
+                // the first line could contain a partial timestamp, so verify it starts with a
+                // null character, otherwise skip
+                if i == 0 && buffer[0] != 0 {
+                    continue;
+                }
+
+                // line could have ended partially through another, so strip anything after `\r`
+                let mut l = l.split('\r');
+                let l = l.next().unwrap().to_string();
+
+                // if on the last chunk, may need to wrap around to the beginning of the buffer
+                if i == lines.len() - 1 && buffer[0] != 0 && !l.ends_with('\n')
+                {
+                    // if wrapping need to strip the \r and anything after it on the wrapped line
+                    let mut wrap = lines.as_slice()[0].split('\r');
+                    let wrap = wrap.next().unwrap();
+                    let s = l + wrap;
+                    full_log.insert(timestamp, (false, s));
+                } else {
+                    // any line but the last
+                    full_log.insert(timestamp, (false, l));
+                }
+            }
+        }
+
+        //
+        // TODO
+        // dont want the btree to grow indefinitly, so cull old entries.
+        // The ringbuf contains at most ringbuf.len lines so  `full_log` just needs to be able to
+        // contain all of them to prevent repeats from being printed
+        // see jira https://rivosinc.atlassian.net/browse/SW-441
+        //while full_log.len() > buffer.len() {
+        //    full_log.pop_first();
+        //}
+
+        // print out all the new entries
+        for (_key, (printed, line)) in full_log.iter_mut() {
+            if !*printed {
+                print!("{}", line);
```

Comment:
I think it would be helpful to print out the timestamps as well.

Reply:
It does!  I updated the inline doc at the top of the file to reflect this.